### ISSUE TITLE
Support Cosma's _link type_ syntax in links

### ISSUE THREAD

**Issue by GuillaumeHM** [Zettlr/Zettlr#5532]
### Checklist

- [X] Zettlr does not possess this feature (if it does, but the feature doesn't work out for you, propose an enhancement instead.)
- [X] Zettlr's current features cannot be used to approximate this behaviour (otherwise, propose an enhancement instead.)
- [X] I have searched the documentation to make sure I did not overlook it.
- [X] I have searched both open and closed issues to see if someone has already proposed this.

### Feature Proposal

I use Zettlr to organize my notes in a Zettlkasten manner and Cosma to search and visualize in. Cosma renders dynamic and pretty graphs from links between notes.

When there are much notes, every way to differentiate them from another is welcome. In order to achieve that, Cosma introduces _link types_, see the manual: [links](https://cosma.arthurperret.fr/user-manual.html#links). So it can colorize arrows between notes following the link type.

It consists to qualify links before the ID : `[[opponent:20201209111626]]`.

It is currently possible to manually write links this way in Zettlr. But it doesn't recognize them as links. So it can't open such target note nor it can differentiate the type from the ID. So it proposes the whole type + ID as IDs in the suggestion popup when typing a new link.

Link type can be considered as a qualifier for the link to the targeted note. This is also very useful in the text in addition to the graph: 
```
[[parent:20201209111626]] An idea note
[[child:20201209111626]] A people note
[[use:20201209111626]] An idea note
[[follows:20201209111626]] A chronological note
...
```
**This participates to the structure description of a whole Zettlekasten note group.** So I think this would bring users a good opportunity to better organize their thoughts.

I suggest to implement the following, from most needed to best featured:
1. Fix the preview and be able to open such type links by just clicking the preview button _Open..._ . For now, this is very painfull since we have to search the note title from the search bar instead). The link preview does show an error message instead of useful information.
2. Have a way to add a type to an existing link which hasn't yet (a pop-up, rising when entering `:` between the link `[[]]` ?). Currently I have to edit the link to add a second ID which contains the desired type then delete its (real) ID part.
3. When typing a new link, do not copy the type along with the ID when selecting a note from suggestion pop-up.
4. Have a checker to guarantee that the type is a known one. Cosma does it but it is painful to come back to the faulty link afterward.
5. Even more nice features like : Given a note, be able to list all notes which it targets, classified by associated type; and so on... But this is also handled by Cosma.

Please let me know what you think about.

### Example Workflow

I see two conceivable worklows:

1. Two times link definition:
    a. User adds a normal link (type is not copied).
    b. User adds a type to its link : He types `:`inside the link (before or after the ID, but not in between), then he selects the desired type name from a dedicated suggestion pop-up.

2. One time, two steps definition:
    a. User start writing a new link, pop-up first suggest the type name (as it comes first). It can be _None_ as default.
    b. Then the same pop-up suggests notes following letters typed by the user, as usual.

### Caveats

I'm not sure there aren't but it seems to be a pretty low probability. The only problem I see could come from the separator character `:`.

### Do you wish to attempt implementing this yourself?

No

### Platforms

- [X] Windows
- [X] macOS
- [X] Linux

### Additional Information

I like do the analogy with HTML and CCS. The first one describes the content and the structure, the second does handle the rendering. In my opinion, Zettlr shall enable writing better structure and content. And Cosma shall bring best viewing and searching utilities.

**Comment by nathanlesage** [comment]
A big potential issue I see here is that this syntax may be incompatible with what Pandoc already offers. @infologie can you say a few words about how cosma can integrate with Pandoc and the broader ecosystem?

Also CCing @sensologica because this issue seems as if it could benefit from some thoughts on UX

**Comment by infologie** [comment]
@nathanlesage The link type syntax we use `[[type:url]]` is inspired by [the category syntax in wiki markup](https://en.wikipedia.org/wiki/Help:Wikitext#Categories), e.g. `[[Help:Contents]]`. Since Pandoc knows how to parse wiki markup, it probably knows how to parse this already. The Pandoc developers have expressed an interest in supporting Markdown wikilinks specifically in reference to note-taking applications. So we could ask them about adding type/category parsing for wikilinks in Markdown input? I hadn't thought about it at all. (Thank you @GuillaumeHM for raising the issue!)

(For additional context, Cosma parses CommonMark + Pandoc style citations, our goal is to eventually parse Pandoc's Markdown but that's far off.)

**Comment by nathanlesage** [comment]
> The Pandoc developers have expressed an interest in supporting Markdown wikilinks specifically in reference to note-taking applications.

Yeah, they added support for this this year! @tarleb do you know how well/unwell broader wikilink support is with Pandoc ...?

Btw @infologie did Cosma's ability increase since the last time we spoke? I'm asking because now I'm wondering whether there's something Zettlr could do in turn regarding cross-app support ...?

**Comment by tarleb** [comment]
Pandoc supports two wikilink flavors, "linktitle before pipe" and "title after pipe". Both are implemented as extensions. Enable the one that fits the format. https://pandoc.org/MANUAL.html#extension-wikilinks_title_after_pipe

**Comment by infologie** [comment]
Cosma uses "title after pipe". Supporting both flavors is on our little roadmap.

@tarleb Am I right in assuming that some Pandoc readers can parse a wiki link with a "category" or "type"? e.g. `[[type:url]]` or `[[type:url|title]]` This is called ["categories"](https://www.mediawiki.org/wiki/Help:Categories) in MediaWiki and ["namespaces"](https://www.dokuwiki.org/wiki:syntax#links) in DokuWiki, for instance. Pandoc has readers for both these languages, hence my assumption. If true, maybe it could be added to the Markdown parser too? This would solve the issue we're discussing here regarding Zettlr and Cosma, while maintaining seamless compatibility with Pandoc's Markdown.

> @nathanlesage: Btw @infologie did Cosma's ability increase since the last time we spoke? I'm asking because now I'm wondering whether there's something Zettlr could do in turn regarding cross-app support ...?

Well, we've continued to add things, but the link "type" syntax is the **only** thing that _breaks_ interoperability when used. So @GuillaumeHM's issue is on point. (Also, we're planning to extend this "type" syntax to citations, e.g. if you want to tag citations with something like CiTO (or your own citation types) you'll be able to write `[type:@key]`.  Contrary to wikilinks types, this won't break compatibility with Pandoc's Markdown because Pandoc treats `[type:@key]` like any citation with a string prefix.)

One idea we could explore would be to add Cosma as an "export format" or "exporter" in Zettlr: when you're in a Zettlr workspace, being able to click on a button that triggers `cosma modelize`, just like exporting a single file triggers `pandoc -o …`. (Cosma is configured via "config.yml" files, which could be integrated as an exporting "asset" in Zettlr.) Our friends from [Stylo](https://stylo-doc.ecrituresnumeriques.ca/en/) (another Markdown academic text editor) showed us that this sort of integration is quite easy to do on a technical level, so we're working with them to turn it into an actual, polished feature. I can share experience when we progress!

**Comment by nathanlesage** [comment]
That sounds great! Adding a new custom exporter would force me to finally rework the exporter, and adding a custom exporter should be straightforward! Let's continue this discussion, because I think Zettlr can greatly benefit from continuous integration with Cosma!

This also reminds me that we should probably at some point add a documentation page on how to get this to work ... 👀

But this year is already full as I will have to finish my PhD in October 🙈

**Comment by GuillaumeHM** [comment]
My end user wish about the idea of a Modelize button for Cosma in Zettlr :
The best would be to have both the button and the generated web page displayed right inside Zettlr ! The ergonomy would be far more better without having to open the page in another window !
It could be implemented throught kind of a plugin, but I don't know if Zettlr has such a sub-system.

**Comment by nathanlesage** [comment]
I do plan on integrating a live preview feature at some point, that is just continuous Pandoc exports and displaying in a webpage inside Zettlr, so that would then be possible -- but that will likely take a long time ...
